Mainnet/STOContract/STOContract/STOContract.cs

```diff
@@ -0,0 +1,1361 @@
+            Transfer(Message.Sender, saleInfo.RefundAmount);
+
+        return true;
+    }
+
+    private void EnsureKycVerified()
+    {
+        var result = Call(MapperAddress, 0, "GetSecondaryAddress", new object[] { Message.Sender });
+
+        if (result.Success && result.ReturnValue is Address identityAddress && identityAddress != Address.Zero)
+        {
+            result = Call(KYCAddress, 0, "GetClaim", new object[] { identityAddress, (uint)3 /*shufti kyc*/ });
+
+            Assert(result.Success && result.ReturnValue != null, "Your KYC is not verified.");
+
+            return;
+        }
+
+        Assert(false, "The address has no mapping.");
+    }
+
+    public bool WithdrawFunds()
+    {
+        Assert(Message.Sender == Owner, "Only contract owner can transfer funds.");
+        Assert(!SaleOpen, "STO is not ended yet.");
+
+        var result = Transfer(this.Owner, Balance);
+
+        return result.Success;
+    }
+
+    public bool WithdrawTokens()
+    {
+        Assert(!IsNonFungibleToken, $"The {nameof(WithdrawTokens)} method is not supported for Non-Fungible Token.");
+        Assert(Message.Sender == Owner, "Only contract owner can transfer tokens.");
+        Assert(!SaleOpen, "STO is not ended yet.");
+
+        var result = Call(TokenAddress, 0, nameof(StandardToken.TransferTo), new object[] { Message.Sender, TokenBalance });
+
+        TokenBalance = 0;
+
+        Assert(result.Success && (bool)result.ReturnValue, "Token transfer failed.");
+
+        return true;
+    }
+
+    private SalePeriod GetCurrentPeriod()
+    {
+        var result = default(SalePeriod);
+
+        foreach (var period in SalePeriods)
+        {
+            if (period.EndBlock >= Block.Number)
+            {
+                result = period;
+                break;
+            }
+        }
+
+        return result;
+    }
+    private void SetPeriods(SalePeriodInput[] periods)
+    {
+        var salePeriods = ConvertSalePeriodInputs(periods);
+
+        SalePeriods = salePeriods;
+        EndBlock = salePeriods[salePeriods.Length - 1].EndBlock;
+    }
+
+    private SalePeriod[] ConvertSalePeriodInputs(SalePeriodInput[] periods)
+    {
+        var result = new SalePeriod[periods.Length];
+        var blockNumber = Block.Number;
+        for (int i = 0; i < periods.Length; i++)
+        {
+            var input = periods[i];
+            blockNumber = checked(blockNumber + input.DurationBlocks);
+            result[i] = new SalePeriod
+            {
+                EndBlock = blockNumber,
+                PricePerToken = input.PricePerToken
+            };
+        }
+
+        return result;
+    }
+
+    private void ValidatePeriods(SalePeriodInput[] periods)
+    {
+        Assert(periods.Length > 0, "Please provide at least 1 sale period");
+
+        foreach (var period in periods)
+        {
+            Assert(period.DurationBlocks > 0, "DurationBlocks should higher than zero");
+        }
+    }
```

Comment:
In reference to `DurationBlocks` validation from this comment, just above: https://github.com/stratisproject/CirrusSmartContracts/pull/31/files#r554280898

can be done here changing this for each to a basic for loop.